utilities/src/parallel.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{boxed::Box, vec::Vec};
+pub struct ExecutionPool<'a, T> {
+    #[cfg(feature = "parallel")]
+    tasks: Vec<Box<dyn 'a + FnOnce() -> T + Send>>,
+    #[cfg(not(feature = "parallel"))]
+    tasks: Vec<Box<dyn 'a + FnOnce() -> T>>,
+}
+
+impl<'a, T> ExecutionPool<'a, T> {
+    pub fn new() -> Self {
+        Self { tasks: Vec::new() }
+    }
+
+    pub fn with_capacity(cap: usize) -> Self {
+        Self {
+            tasks: Vec::with_capacity(cap),
+        }
+    }
+
+    #[cfg(feature = "parallel")]
+    pub fn add_task<F: 'a + FnOnce() -> T + Send>(&mut self, f: F) {
```

Comment:
```suggestion
    pub fn add_job<F: 'a + FnOnce() -> T + Send>(&mut self, f: F) {
```
I'd change this name in order not to mix it up with asynchronous tasks.

Reply:
Done, thanks!